tla-io/src/main/scala/at/forsyte/apalache/io/ConfigManager.scala

```diff
@@ -0,0 +1,126 @@
+package at.forsyte.apalache.io
+
+import pureconfig._
+import pureconfig.generic.auto._
+import java.io.{File}
+import java.nio.file.{Path, Files, Paths}
+
+object Converters {
+
+  private def expandedFilePath(s: String): Path = {
+    Paths.get(if (s.startsWith("~")) s.replaceFirst("~", System.getProperty("user.home")) else s)
+  }
+
+  // Value class to allow adding a new implicit for the File
+  case class ExpandedFile(val file: File) extends AnyVal
+
+  object ExpandedFile {
+    def apply(f: File): ExpandedFile = {
+      new ExpandedFile(expandedFilePath(f.toString()).toFile())
+    }
+  }
+
+  // Value class to allow adding a new implicit for Path
+  case class ExpandedPath(val path: Path) extends AnyVal
+
+  object ExpandedPath {
+    def apply(f: Path): ExpandedPath = {
+      new ExpandedPath(expandedFilePath(f.toString()))
+    }
+  }
+
+  // Briniging these implicits in scope lets us override the existing
+  // file deserialization behavior, so we get path expansion in all configured
+  // paths
+  implicit def expandedFileConfigReader: ConfigReader[ExpandedFile] =
+    ConfigReader[File].map(ExpandedFile.apply)
+
+  implicit def expandedPathConfigReader: ConfigReader[ExpandedPath] =
+    ConfigReader[Path].map(ExpandedPath.apply)
+}
+
+/** The configuration values that can be overriden based on CLI arguments */
+trait CliConfig {
+
+  /** Input file */
+  def file: File
+  def outDir: Option[File]
+  def runDir: Option[File]
+  def writeIntermediate: Option[Boolean]
+  def profiling: Option[Boolean]
+  def configFile: Option[File]
+}
+
+/** The application's configurable values, along with their base defaults */
+case class ApalacheConfig(
+    file: Option[File] = None, outDir: File = new File(System.getProperty("user.dir"), "_apalache-out"),
+    runDir: Option[File] = None, configFile: Option[File] = None, writeIntermediate: Boolean = false,
+    profiling: Boolean = false
+)
+
+case class ConfigManager(cmd: CliConfig) {
+  val TLA_PLUS_DIR = ".tlaplus"
+  val APALACHE_CFG = "apalache.cfg"
+  val DOT_APALACHE_CFG = "." + APALACHE_CFG
```

Comment:
Wouldn't these be better suited for a companion object?

Reply:
I don't know :)

What's the advantage of putting it it in the companion object?

I do think these should private tho. I'll at least change that.

I can't find anything in the Scala style guide that says where fields like this should be declared. Could you point me towards something that explains why putting such things in companion object might be preferable?

My own feeling is that -- in this case -- moving these into the companion object it would negatively impact readability: these  values are only accessed in the class methods, and don't show up anywhere in the one method in the companion object. Moving them would then require readers to jump back and forth between the class and the companion object to see what means what.

Reply:
I think we generally have the convention across the codebase that any `public static X`-type value goes in a companion object, so it can be accessed without creating a fresh instance. 
This page quotes the "Scala Cookbook": https://alvinalexander.com/scala/how-to-static-members-in-scala-companion-objects-fields-methods/

Reply:
Does the fact that these are (now) private and *not* meant to be accessed outside of the class at all impact your feelings on this? (I.e., they are never supposed to be accessed at all, whether by creating instance or not).

I still don't see there anything that explains a benefit in this case, and the downside on readability (and abstraction) seem pretty evident to me. I don't want to buck our conventions, but I'd rather there be a good reason if I'm going to negatively impact readability.

Reply:
A bit more concretely, that recipe is for the case where

> You want to create a Scala class that has both instance methods and static methods, but unlike Java, Scala does not have a static keyword.

afaict, that's not what we're actually after here.